File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long credit_number = get_long("Number:");
    string card_type;
    int length = 0;
    long poww(long a, int b);
    int digit;
    int summ = 0;

    if ((340000000000000 <= credit_number && credit_number < 350000000000000) ||
        (370000000000000 <= credit_number && credit_number < 380000000000000))
    {
        card_type = "AMEX\n";
        length = 15;
    }
    else if (5100000000000000 <= credit_number && credit_number < 5600000000000000)
    {
        card_type = "MASTERCARD\n";
        length = 16;
    }
    else if (4000000000000000 <= credit_number && credit_number < 5000000000000000)
    {
        card_type = "VISA\n";
        length = 16;
    }
    else if (4000000000000 <= credit_number && credit_number < 5000000000000)
    {
        card_type = "VISA\n";
        length = 13;
    }
    else
    {
        printf("INVALID\n");
        return (0);
    }

    for (int i = 0; i < length; i++)
    {

        long dosens = poww(10, i);

        if (i % 2 == 0)
        {
            digit = (credit_number % dosens) / (dosens / 10);
        }
        else
        {
            digit = 2 * (credit_number % dosens) / (dosens / 10);
            digit = 2 * digit;

            if (digit > 9)
            {
                digit = (digit % 10) + 1;
            }

        }

        summ += digit;
        printf("%i\n", digit);
    }

    if (summ % 10 == 0)
    {
        printf("%s", card_type);
    }
    else
    {
        printf("INVALID\n");
    }
}

long poww(long a, int b)
{
    long c = a;
    for (int i = 0; i < b; i++)
    {
        a *= c;
    }
    return (a);
}

File: _cs50x/sentimental-credit/credit.py
from sys import exit
while (True):
    try:
        number = str(int(input('Number: ')))
        break
    except ValueError:
        print("INVALID")

digits = len(number)

if digits == 15 and number[0] == "3" and (number[1] == "4" or number[1] == "7"):
    card_type = "AMEX"
elif digits == 16 and number[0] == "5" and (int(number[1]) in range(1, 6)):
    card_type = "MASTERCARD"
elif ((digits == 13) or (digits == 16)) and (number[0] == "4"):
    card_type = "VISA"
else:
    print("INVALID")
    exit()


check_sum = 0
for i in range(digits):
    index = digits - i - 1
    if i % 2:
        j = int(number[index])*2
        if j > 9:
            j = j - 9
        check_sum += j
    else:
        check_sum += int(number[index])
        

if check_sum % 10 == 0:
    print(card_type)
else:
    print("INVALID")
